### PR TITLE
[api-extractor] Add options to include forgotten exports in API report and doc model files

### PR DESCRIPTION
**Pull request by zelliott**
## Summary

Fixes https://github.com/microsoft/rushstack/issues/3513.

## Details

* Adds `apiReport.includeForgottenExports` and `docModel.includeForgottenExports` configs to `api-extractor.json` that control whether forgotten exports are included in the API report and doc model files, respectively.
* Adds a new `ApiExportedMixin` mixin to api-extractor-model to allow API items to store whether they are exported or not from their container module (i.e. entry point or namespace). This is necessary to build the proper canonical reference (i.e. with or without a `~` navigation step) from the `.api.json`. Note that this mixin doesn't actually serialize the new `isExported` property to the `.api.json` because this info is already present in the canonical reference, thus no `.api.json` change needed.
* Reworks `CollectionEntity` and the parts of the `Collector` that construct them to change the definitions of "consumable" and "exported" as previously defined by api-extractor. More on this below.

### Consumable vs exported

Before this PR, collector entities could be exported and/or consumable. An exported entity was a top-level export, like:

```
/** @public */
export declare class A {}
```

whereas a consumable entity was either a top-level export OR an entity exported from an `AstNamespaceImport`. An example of the latter might be:

```
declare class B {}

/** @public */
export declare namespace n {
  export {
    B
  }
}
```

This previous definition of consumable is incomplete and misleading for a few reasons:

1. It doesn't actually care whether the `AstNamespaceImport` `n` is exported. This is a bug. This PR fixes that bug.
2. It only applies to `AstNamespaceImport` namespaces, not normal `AstSymbol` namespaces. There's fundamentally no reason why the two should be treated differently. This also leads to bugs. This PR doesn't update `AstSymbol` namespaces to be treated similarly to `AstNamespaceImport`s, but reworks `CollectionEntity` and the parts of the `Collector` that construct them in order to pave the way for a future PR to do this.

This PR changes the definition of consumable to the following: **A consumable entity is either a top-level export or an entity exported from a consumable parent namespace.**

This PR also updates the definition of exported. The previous definition is not particularly useful today (almost all code paths use `consumable` instead of `exported`). Instead, we need to know whether an entity is exported with respect to its parent module, for the purposes of writing the proper navigation step in api-extractor-model. Thus, this PR updates the definition of exported to the following: **An exported entity is an entity exported from its parent API item.**

### Ambiguous IDs

As discussed in https://github.com/microsoft/rushstack/issues/3513, unexported API items will encounter an instance of the "ae-ambiguous-ids" problem. In short, it's possible for there to be multiple unexported API items with the same name that all need to be written to the `.api.json`. Consider the following:

```
// foo.ts

type SomeType = number;
export function foo(): SomeType { return 5; }

// bar.ts

type SomeType = number;
export function bar(): SomeType { return 5; }

// index.ts (entry point)

export { foo } from './foo';
export { bar } from './bar';
```

`foo.ts:SomeType` and `bar.ts:SomeType` are both forgotten exports and will be written to the doc model. if `docModel.includeForgottenExports` is enabled. What names and canonical references do we give these items?

This PR reuses the existing logic in `Collector._makeUniqueNames` to assign a unique `nameForEmit` for each item. In practice, one `SomeType` will keep it's name, and the other will be de-duped to `SomeType_2`. These unique names will be used when writing both the items' names and canonical references to the doc model. The `includeForgottenExports` test case tests this scenario.

There are a few important things to note about this:

1. *This is not a new problem.* This existing behavior exists today if an exported item's name conflicts with a global name. The exported item will be renamed (e.g. see the `ambientNameConflict2` test case, `Date` and `Date_2`). This PR uses this same solution.
2. These names are not stable, names can be shuffled if new items with the same name are introduced. But this can be solved in the future with the `@label` tag as discussed in https://github.com/microsoft/rushstack/issues/3513.

### Known bugs/missing features

This is a list of some known bugs and missing features associated with this PR. If anything in this list should block this PR being submitted, please let me know!

1. Canonical references in reference tokens (constructed via `DeclarationReferenceGenerator`) use the item's original name, not it's unique `nameForEmit` name. I didn't include this fix in this PR because this is an existing bug.
2. Forgotten exports within consumable namespaces are included in the `.api.json` and `.api.md` even if the `includeForgottenExports` setting is off (https://github.com/microsoft/rushstack/pull/3552#discussion_r930244544). I didn't include this fix in this PR because this is an existing bug.
3. `@link` and `@inheritDoc` references to unexported API items cannot be resolved. I didn't include this fix in this PR because it seemed like a a relatively minor missing feature and easy to fix as a follow-on task.

## How it was tested

I added a new test to api-extractor-scenarios and manually validated that stuff looked correct.

### CODE REVIEW

File: apps/api-extractor/src/collector/Collector.ts [hunk 1]
```diff
@@ -425,50 +424,54 @@ export class Collector {
...
   }
 
-  private _createEntityForIndirectReferences(
-    astEntity: AstEntity,
-    alreadySeenAstEntities: Set<AstEntity>
-  ): void {
-    if (alreadySeenAstEntities.has(astEntity)) {
-      return;
-    }
+  private _recursivelyCreateEntities(astEntity: AstEntity, alreadySeenAstEntities: Set<AstEntity>): void {
```
Comment (zelliott) [inline apps/api-extractor/src/collector/Collector.ts:438]: I renamed this method to `_recursivelyCreateEntities` because the previous name was inaccurate. The previous name implied that we were only creating entities for indirect references, when in fact we also create entities for things within `AstNamespaceImport` entities.

File: apps/api-extractor/src/collector/Collector.ts [hunk 2]
```diff
@@ -425,50 +424,54 @@ export class Collector {
...
-            // We only create collector entities for root-level symbols.
-            // For example, if a symbols is nested inside a namespace, only the root-level namespace
-            // get a collector entity
+            // We only create collector entities for root-level symbols. For example, if a symbol is
+            // nested inside a namespace, only the namespace gets a collector entity. Note that this
+            // is not true for AstNamespaceImports below.
             if (referencedAstEntity.parentAstSymbol === undefined) {
-              this._createCollectorEntity(referencedAstEntity, undefined);
+              this._createCollectorEntity(referencedAstEntity);
             }
```
Comment (zelliott) [inline apps/api-extractor/src/collector/Collector.ts:451]: Long-term, to fix bugs like https://github.com/microsoft/rushstack/issues/2808 as well as track whether items within namespaces are exported, we'll need to create collector entities for nested symbols. For example:

```
export declare namespace Foo {
  declare class Bar {}
  export declare class Baz extends Bar {}
}
```

Right now, `Foo`, `Bar`, and `Baz` will all be marked as exported, because there's only one collector entity for all three API items (the entity for `Foo`). In reality, only `Foo` and `Baz` should be marked as exported, and `Bar` and `Baz` should have their own collector entities.

I think this will need to be in a separate PR though, as it's a larger change.

File: apps/api-extractor/src/collector/Collector.ts [hunk 3]
```diff
@@ -826,10 +829,14 @@ export class Collector {
               astSymbol
             );
           }
+
+          // All internal, exported APIs default to public if no release tag is specified.
+          options.effectiveReleaseTag = ReleaseTag.Public;
         }
+      } else {
+        // All external APIs default to public if no release tag is specified.
+        options.effectiveReleaseTag = ReleaseTag.Public;
```
Comment (zelliott) [inline apps/api-extractor/src/collector/Collector.ts:838]: This change is to prevent unexported items included in an API report from being assigned the `@public` release tag if none is specified.

File: apps/api-extractor/src/enhancers/ValidationEnhancer.ts [hunk 4]
```diff
@@ -210,9 +210,8 @@ export class ValidationEnhancer {
           continue;
         }
 
-        localName = rootSymbol.localName;
-
         collectorEntity = collector.tryGetCollectorEntity(rootSymbol);
+        localName = collectorEntity?.nameForEmit || rootSymbol.localName;
```
Comment (zelliott) [inline apps/api-extractor/src/enhancers/ValidationEnhancer.ts:214]: This change and the one below is a nit bug fix unrelated to this PR (I noticed it in the test case I added). See https://github.com/microsoft/rushstack/pull/3552/files#diff-f6ecec4f81407d8e726a025aa7aef71db73f5992650ddccd959c355d6a02e31bR7.

File: build-tests/api-extractor-lib2-test/etc/api-extractor-lib2-test.api.md [hunk 5]
```diff
@@ -7,7 +7,6 @@
 // @public (undocumented)
 class DefaultClass {
 }
-
```
Comment (zelliott) [inline build-tests/api-extractor-lib2-test/etc/api-extractor-lib2-test.api.md:10]: I'm not sure why this whitespace is changing here and in some other files, but I believe these changes are wanted. 🤷  It makes API report files more consistent.

File: build-tests/api-extractor-scenarios/etc/exportImportStarAs2/api-extractor-scenarios.api.md [hunk 6]
```diff
@@ -9,10 +9,6 @@
 // @public (undocumented)
 function exportedApi(): forgottenNs.ForgottenClass;
 
-// @public (undocumented)
-class ForgottenClass {
-}
-
```
Comment (zelliott) [inline build-tests/api-extractor-scenarios/etc/exportImportStarAs2/api-extractor-scenarios.api.md:15]: I think we want this change. `ForgottenClass` is exported by `forgottenNs` **but** `forgottenNs` isn't actually exported itself. Thus, `ForgottenClass` isn't exported and thus shouldn't appear in the API report. I believe previously `ForgottenClass` was marked as consumable when it in fact was not (this was a bug).

File: build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.json [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+          "docComment": "/**\n * @public\n */\n",
+          "excerptTokens": [
+            {
+              "kind": "Content",
+              "text": "export declare function anotherFunction(): "
+            },
+            {
+              "kind": "Reference",
+              "text": "DuplicateName",
+              "canonicalReference": "api-extractor-scenarios!~DuplicateName:type"
```
Comment (zelliott) [inline build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.json:212]: This should be `DuplicateName_2`, but the declaration reference builder doesn't use `CollectorEntity`s and thus doesn't know about any renaming that occurred. This is an existing bug however and should be fixed in a separate PR.

File: build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
+## API Report File for "api-extractor-scenarios"
+
+> Do not edit this file. It is a report generated by [API Extractor](https://api-extractor.com/).
+
+```ts
+
+// Warning: (ae-unresolved-inheritdoc-reference) The @inheritDoc reference could not be resolved: The package "api-extractor-scenarios" does not have an export "ForgottenExport"
+//
```
Comment (zelliott) [inline build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.md:8]: **TODO:** This is a bug that I need to fix. It looks like `@inheritDoc` (and probably also `@link`) references expect to resolve to exported symbols.

File: libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+      options: Partial<IApiExportedMixinOptions>,
+      context: DeserializerContext,
+      jsonObject: IApiExportedMixinJson
+    ): void {
+      baseClass.onDeserializeInto(options, context, jsonObject);
+
+      const declarationReference: DeclarationReference = DeclarationReference.parse(
+        jsonObject.canonicalReference
+      );
+      options.isExported = declarationReference.navigation === Navigation.Exports;
```
Comment (octogonz) [inline libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts:82]: This seems potentially brittle, since the canonical reference algorithm is fairly complex, and the declaration reference parser is experimental. Since we're relying on it, maybe **ApiModelGenerator.ts** can validate that `declarationReference.navigation === Navigation.Exports` when it is writing the JSON file.

File: libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+export interface IApiExportedMixinJson extends IApiItemJson {
+  isExported: boolean;
+}
+
+const _isExported: unique symbol = Symbol('ApiExportedMixin._isExported');
+
+/**
+ * The mixin base class for API items that can be exported.
```
Comment (octogonz) [inline libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts:23]: What qualifies as "exported"?

**src/index.ts** (entry point for API Extractor)
```ts
export { C } from './file';
```

**src/file.ts**
```ts
// C is "exported"
export class C { }

// D is exported by file.ts, but it is not exported by the package.
// Is D "exported"?
export class D { }

// n is not "exported"
namespace n {
  // Is f() "exported"?
  export function f(): void { }
}
```

Maybe we need a different jargon, for example "importable"?

File: libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts (hunk 9 above)
Comment (zelliott) [reply libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts:82]: I can understand why this code is brittle if it's relying upon experimental code. But I don't quite follow the second sentence - do you mind clarifying or pointing me to where exactly this validation would occur and how it would make this less brittle?

File: libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts (hunk 10 above)
Comment (zelliott) [reply libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts:23]: > What qualifies as "exported"?

An item is exported if it is exported from the package, that is, exported from the package's single entry point. Local exports (e.g. `D`, `n.f`) are not exported from the package and thus are not "exported". They might still be exported from a local file or a local non-exported namespace.

Another option is "consumable", but that doesn't quite work because if you have:

```
export class A {
  public prop: number;
}
```

I think most would interpret `A.prop` as consumable, whereas in our usage it's not the case.

Another pro for `ApiExportedMixin` over `ApiImportableMixin` is that it aligns with the `Navigation.Exports` navigation step used by TSDoc (i.e. an item has `isExported` set to true IFF its navigation step is `Navigation.Exports`).

@wcauchois do you have any opinions?

Comment (octogonz) [reply libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts:23]: I understand. However for a novice, it might be challenging to understand under what conditions `isExported=false` for a declaration that conspicuously includes the word `export`.  

On the other hand, an API consumer can import `C` using `import { C } from 'library'`, but they cannot use `import` with any of the other examples `A.prop`, `D`, `n`, `n.f`.  "Importable" might not be the best terminology, but maybe it shows that this concept could be simplified.

File: libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts (hunk 9 above)
Comment (octogonz) [reply libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts:82]: I was thinking of this code, since `getDeclarationReferenceForIdentifier()` is a pretty complicated analysis:

https://github.com/microsoft/rushstack/blob/0e2502ea1e4aec2d0cbf912ed5712b8f05b0a551/apps/api-extractor/src/generators/ExcerptBuilder.ts#L150-L155

But it turns out this is not used for `ApiItem.canonicalReference`, which is built directly from API Extractor's tree by `buildCanonicalReference()`.  I think that addresses my concern. 👍

File: libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts (hunk 10 above)
Comment (zelliott) [reply libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts:23]: So ultimately, the purpose of `ApiExportedMixin` and the `ApiExportedMixin.isExported` property is two-fold:

#### 1. Allows api-extractor-model to construct the correct reference for a given API item.

api-extractor-model needs to know whether to write a `Navigation.Exports` or `Navigation.Local` step when constructing an API item's reference. Otherwise, the reference constructed for an item in a reference token won't match the reference constructed for an item in the doc model, and thus reference resolution won't work. Today, the navigation step chosen when constructing the reference in a reference token is done by the `DeclarationReferenceGenerator._getNavigationToSymbol` method, while the navigation step chosen when constructing the reference for an item is hardcoded in api-extractor-model in the item's `buildCanonicalReference` method (typically to `Navigation.Exports` for the top-level items we care about).

This hardcoding of `Navigation.Exports` worked in the past because we only had exported items in the doc model (this actually isn't quite true, unexported items in namespaces can still be included in the doc model, but we can ignore this for now, this is a bug AFAICT). But now that we're including unexported items in the doc model under certain conditions, we need a way to set the appropriate navigation step.

Whatever mixin/logic we use to do so, it needs to as closely match the current logic for setting navigation steps in `DeclarationReferenceGenerator._getNavigationToSymbol`. Whatever definition of "exported" is being used there to set `Navigation.Exports` should be the same definition that is used in our new mixin/logic.

#### 2. Allows a documentation tool such as api-documenter to show on an API item's documentation page whether or not that item is "exported" (for some definition of the word).

It's useful for users to know whether an API item is "consumable/importable" or "exported from the current package". In https://github.com/microsoft/rushstack/issues/3513#issuecomment-1191791900, @wcauchois proposed including or omitting the `export` keyword in the API item's signature to communicate this. The problem here is that as @octogonz stated it's possible for an API item to be exported from a file but not actually exported from the package's single entry point. Should those items get `export` keywords in their signatures?

This proposal might also be too subtle for users (many users might miss this). I'm envisioning a more prominent flag or blurb somewhere on the API item's page that indicates whether someone can import/consume it.

Comment (zelliott) [reply libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts:23]: The problem with `ApiImportableMixin` is the following setup:

```
/** @public */
export namespace n {
  export class Foo {}
}
```

The declaration reference for `Foo` should be something like `package!n.Foo:class`. `n` uses the `Navigation.Exports` step and `Foo` uses the `Navigation.Exports` step. However, `Foo` uses the `Navigation.Exports` step because it's exported from `n`, not because it's importable (which it isn't, you can't import `Foo`).

So I think I misspoke above when I said:

> > What qualifies as "exported"?
>
> An item is exported if it is exported from the package, that is, exported from the package's single entry point.

Instead, I'm trying to use the same definition of "exported" as is used by the `DeclarationReferenceGenerator._getNavigationToSymbol` when determining what navigation step to use. The definition that I believe is being used there is: _"An item is exported if it's exported from the package or exported from a namespace within the package."_ So for a scenario like this:

```
namespace n {
  export class Foo {}
}
```

`n` is not exported, but `Foo` is exported. Whether or not an item is exported is determined with respect to its parent "module" (I think that's right?).

Note that this definition of "exported" is less useful to surface on a documentation site. A user doesn't care whether or not `Foo` is exported from `n`, they care whether `Foo` is "consumable". Instead, we'd probably want to expose a function in api-extractor-model that you can use to determine whether an item is consumable by checking if it and its entire parent tree is exported.

Comment (zelliott) [reply libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts:23]: > > What qualifies as "exported"?
>
> An item is exported if it is exported from the package, that is, exported from the package's single entry point. Local exports (e.g. D, n.f) are not exported from the package and thus are not "exported". They might still be exported from a local file or a local non-exported namespace.

I now realize that the definition I provided here and the one I'm using in this PR are not consistent with what the `Navigation.Exports` step represents. The two need to be consistent, because I've implemented that an item has a `Navigation.Exports` step if it is exported.

In `DeclarationReferenceGenerator._getNavigationToSymbol`, an item seems to have a `Navigation.Exports` step if it is exported from its parent module (i.e. either the package entry point or a local namespace). Whereas in this PR, I've implemented that an item has a `Navigation.Exports` step if `ApiExportedMixin.isExported` is true, and `ApiExportedMixin.isExported` is true if the item is exported from the package. So I'll need to iterate on this a bit... 🤔 

Perhaps we should return to the previous words _consumable_ and _exported_, albeit with different definitions from before:

### Before this PR

* *Exported:* A top-level export.
* *Consumable:* Either a top-level export or exported from an exported namespace.

### After

* *Exported:* An item exported from its parent module (i.e. either package entry point or local namespace).
* *Consumable:* Same as above.

So in the examples below:

```
export class A {}
class B {}

export namespace n1 {
  export class C {}
  class D {}
}

namespace n2 {
  export class E {}
  class F {}
}
```

### Before this PR

| Item | Exported | Consumable |
| --- | --- | --- |
| A | Yes | Yes |
| B | No | No |
| C | No | Yes |
| D | No | No |
| E | No | No |
| F | No | No |

### After

| Item | Exported | Consumable |
| --- | --- | --- |
| A | Yes | Yes |
| B | No | No |
| C | Yes | Yes |
| D | No | No |
| E | Yes | No |
| F | No | No |

With these new definitions, we can write the appropriate navigation step.

Comment (zelliott) [reply libraries/api-extractor-model/src/mixins/ApiExportedMixin.ts:23]: I updated the PR to use these new definitions, thus addressing the inconsistency between what `ApiExportedMixin.isExported` means and what `Navigation.Exports` means. Also updated the original PR description.

File: build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.json [hunk 11]
```diff
@@ -0,0 +1,550 @@
...
+              "docComment": "",
+              "excerptTokens": [
+                {
+                  "kind": "Content",
+                  "text": "export function someFunction3(): "
+                },
+                {
+                  "kind": "Reference",
+                  "text": "ForgottenExport3",
+                  "canonicalReference": "api-extractor-scenarios!~ForgottenExport3:class"
```
Comment (zelliott) [inline build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.json:529]: Notice that the canonical reference here is missing the namespace. I believe this is an existing bug in `DeclarationReferenceGenerator`.

Comment (zelliott) [reply build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.json:529]: Opened a fix at https://github.com/microsoft/rushstack/pull/3566.

File: build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.json [hunk 12]
```diff
@@ -0,0 +1,638 @@
...
+              "text": "declare namespace ForgottenExport4 "
+            }
+          ],
+          "releaseTag": "None",
+          "name": "ForgottenExport4",
+          "preserveMemberOrder": false,
+          "members": [
+            {
+              "kind": "Class",
+              "canonicalReference": "api-extractor-scenarios!~ForgottenExport4~ForgottenExport5:class",
```
Comment (zelliott) [inline build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.json:360]: This canonical reference should be `api-extractor-scenarios!~ForgottenExport4.ForgottenExport5:class`, because `ForgottenExport5` is exported from `ForgottenExport4`. However, unfortunately `ApiExportedMixin.isExported` is `false` for `ForgottenExport5`, because `ForgottenExport5` determines its exported state from `ForgottenExport4`'s `CollectorEntity` (i.e. it doesn't have its own `CollectorEntity`).

This is fundamentally an instance of the pre-existing bug 2 mentioned in the PR description.

File: build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.json [hunk 13]
```diff
@@ -0,0 +1,638 @@
...
+          "docComment": "/**\n * @public\n */\n",
+          "excerptTokens": [
+            {
+              "kind": "Content",
+              "text": "export declare function someFunction2(): "
+            },
+            {
+              "kind": "Reference",
+              "text": "DuplicateName",
+              "canonicalReference": "api-extractor-scenarios!~DuplicateName:type"
```
Comment (zelliott) [inline build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.json:443]: This canonical reference should be `api-extractor-scenarios!~DuplicateName_2:type`. This is the pre-existing bug 1 mentioned in the PR description.

File: build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.json [hunk 14]
```diff
@@ -0,0 +1,638 @@
...
+              "text": "export declare function someFunction5(): "
+            },
+            {
+              "kind": "Content",
+              "text": "internal2."
+            },
+            {
+              "kind": "Reference",
+              "text": "ForgottenExport6",
+              "canonicalReference": "api-extractor-scenarios!ForgottenExport6:class"
```
Comment (zelliott) [inline build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.json:503]: Observe that the canonical reference for `ForgottenExport6` as determined from the `.api.json` above is `api-extractor-scenarios!internal2.ForgottenExport6:class`. This is an existing bug with `AstNamespaceImport`s, discussed a bit in https://github.com/microsoft/rushstack/pull/3584#discussion_r944523487.

File: build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.md [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+class ForgottenExport1 {
+    constructor();
+    // Warning: (ae-forgotten-export) The symbol "ForgottenExport2" needs to be exported by the entry point index.d.ts
+    //
+    // (undocumented)
+    prop?: ForgottenExport2;
+}
+
+// Warning: (ae-unresolved-inheritdoc-reference) The @inheritDoc reference could not be resolved: The package "api-extractor-scenarios" does not have an export "ForgottenExport1"
```
Comment (zelliott) [inline build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.md:27]: This is bug 3 mentioned in the PR description.

File: apps/api-extractor/src/schemas/api-extractor-template.json [hunk 16]
```diff
@@ -175,7 +176,16 @@
...
+    // "reportTempFolder": "<projectFolder>/temp/",
+
+    /**
+     * Whether "forgotten exports" should be included in the API report file. Forgotten exports are declarations
+     * flagged with `ae-forgotten-export` warnings. See https://api-extractor.com/pages/messages/ae-forgotten-export/ to
+     * learn more.
+     *
+     * DEFAULT VALUE: "false"
+     */
+    // includeForgottenExports?: boolean
```
Comment (octogonz) [inline apps/api-extractor/src/schemas/api-extractor-template.json:188]: `?` is not valid here because this is a JSON file

Comment (octogonz) [reply apps/api-extractor/src/schemas/api-extractor-template.json:188]: Same issue for includeForgottenExports

File: apps/api-extractor/src/schemas/api-extractor-template.json [hunk 17]
```diff
@@ -175,7 +176,16 @@
      * SUPPORTED TOKENS: <projectFolder>, <packageName>, <unscopedPackageName>
      * DEFAULT VALUE: "<projectFolder>/temp/"
      */
-    // "reportTempFolder": "<projectFolder>/temp/"
+    // "reportTempFolder": "<projectFolder>/temp/",
+
+    /**
+     * Whether "forgotten exports" should be included in the API report file. Forgotten exports are declarations
+     * flagged with `ae-forgotten-export` warnings. See https://api-extractor.com/pages/messages/ae-forgotten-export/ to
```
Comment (octogonz) [inline apps/api-extractor/src/schemas/api-extractor-template.json:183]: We have historically avoided using `` ` `` or other markdown inside JSON config file comments.  Editors don't highlight it, and it's unclear whether TSDoc can be used there.  But we do use `/**` comment framing, which I suppose is confusing.

🤔 I suppose we should either (1) decide to adopt TSDoc in JSON config files, or else (2) replace `/**` with `/*` to clarify that they are plain text comments.  Let me know your opinion. I'll ask some other maintainers as well.

Comment (octogonz) [reply apps/api-extractor/src/schemas/api-extractor-template.json:183]: Consensus is that we should allow TSDoc in config files. Don't worry about it too much in this PR -- we'll make a separate PR to tune up the `/**` syntax in config files

**Comment by octogonz** [comment]
👍 I tried running this on a bunch of monorepo projects with `apiReport.includeForgottenExports=false` and `docModel.includeForgottenExports=false` and the output was unchanged. It doesn't prove the feature works, but it gives some confidence that the PR didn't introduce any regressions.

File: apps/api-extractor/src/collector/Collector.ts (hunk 3 above)
Comment (octogonz) [reply apps/api-extractor/src/collector/Collector.ts:838]: After some discussion, I think this is not right:  The effective release tag must always be something other than `ReleaseTag.None`.  

The simplest approach would be to handle forgotten exports like any other API item:
- API Extractor complains if it doesn't have a release tag (because for example .d.ts trimming is not going to produce good results if developers don't think carefully about what is/isn't public)
- If you disable the warnings, then everything defaults to `@public` (because in this case, the developer presumably doesn't care very much about trimming or release tags)

File: build-tests/api-extractor-lib2-test/etc/api-extractor-lib2-test.api.md (hunk 5 above)
Comment (octogonz) [reply build-tests/api-extractor-lib2-test/etc/api-extractor-lib2-test.api.md:10]: Not sure. I ran `common\autoinstallers\rush-prettier\node_modules\.bin\prettier -w .` in the repo root folder for your branch, and it didn't produce any diff.

File: build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.json (hunk 7 above)
Comment (octogonz) [reply build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.json]: Agreed

File: build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.md (hunk 8 above)
Comment (zelliott) [reply build-tests/api-extractor-scenarios/etc/includeForgottenExports/api-extractor-scenarios.api.md]: I ended up deciding not to fix this, see the "Known bugs/missing features" section in the PR description.

File: apps/api-extractor/src/collector/Collector.ts (hunk 3 above)
Comment (zelliott) [reply apps/api-extractor/src/collector/Collector.ts]: Updated this PR to match the behavior you described above. However, I was thinking a bit more about forgotten exports and release tags and sent you a message on Zulip, reproduced below:

> We decided yesterday that the missing release tag warning should only be logged if includeForgottenExports is on. I'm wondering now if the missing release tag warning should be logged regardless of whether that config is on. Why? Because .d.ts trimming will produce better results if forgotten exports have release tags.
>
> Maybe the original thinking was that a developer will already get an ae-forgotten-export warning for these items, and so there's already "some warning" indicating that these items should be exported (and thus also should have release tags). But (1) what if the developer turned off that warning and (2) even if the developer didn't, it still might be useful to report both warnings (otherwise a developer might slap on an export, re-run API Extractor, and then just hit another warning right after, as they forgot to also add a release tag).

Comment (octogonz) [reply apps/api-extractor/src/collector/Collector.ts]: Agreed, but I think we could implement that extra validation in a separate PR.

**Review by octogonz — APPROVED** [review]

**Comment by octogonz** [comment]
🚀 This feature was released with API Extractor 7.30.0
